WcaOnRails/app/models/abstract_cached_model.rb

```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+class AbstractCachedModel < ActiveRecord::Base
```

Comment:
I just pushed a version using rails concern that achieve the same feature :)

I must admit I'm not very familiar with the Ruby ways of doing things. To me it seems very not intuitive to disconnect the caching system from the underlying data structure it manipulates.
In this particular case `include Cachable` doesn't make sense if it's not in a model inheriting from `ActiveRecord::Base` (it will simple crash on `all.index_by`), which is why extending `ActiveRecord::Base` feels like "the" thing to do for me.

But again I'm biased because usually when I write code it has to compile, so writing this kind of stuff feels weird from the beginning.
So if using Mixins seems like a better Ruby code I'm totally fine with doing it this way :)

Regarding the solution used, I really like 2 and 3. 3 sounds like the best option if we care about cache invalidation.
As you said it would solve a problem we may not really have, given the very very low chance that we may have to change one of the relevant tables without restarting the app, I'd vote for using 2.